CriteoPublisherSdk/Sources/Util/CR_URLResolver.m

```diff
@@ -0,0 +1,121 @@
+//
+//  CR_URLResolver.m
+//  CriteoPublisherSdk
+//
+//  Copyright © 2018-2020 Criteo. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "CR_URLResolver.h"
+#import "Logging.h"
+#import "CR_DeviceInfo.h"
+#import "CR_URLRequest.h"
+
+@interface CR_URLResolution ()
+@property(nonatomic, strong) NSURL *URL;
++ (instancetype)resolutionWithType:(CR_URLResolutionType)type;
++ (instancetype)resolutionError:(NSError *)error;
+@end
+
+@interface CR_URLResolver ()
+@end
+
+@interface CR_URLResolver () <NSURLSessionDataDelegate>
+@property(nonatomic, strong, readonly) NSURLSession *urlSession;
+@property(nonatomic, copy) CR_URLResolutionHandler resolution;
+@end
+
+@implementation CR_URLResolver
+
+#pragma mark - Lifecycle
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _urlSession =
```

Comment:
Shouldn't you share this session with the one used in the NetworkManager class ?
You might declare the session in the dependency provider and provide it to both NetworkManager and URLResolver.

Reply:
Ideally yes, session should be shared to leverage cache. This is a deliberate choice and tradeoff for simplification because a session has only one delegate for all the tasks, it become rapidly complex. And on the other hand, from my understanding, when reusing the session, in the url resolver context, the benefits are small.

Reply:
Ok then